### ISSUE TITLE
Active Directory Login requires a little more work than the example shows.

### ISSUE THREAD

**Issue by nathanbrizzee** [feathersjs-ecosystem/feathers-authentication-ldap#33]
Thanks for all the work on this module.  For me, I had to make some modifications to the example to get it to work properly.

Following the example, I would get the error:
error: OperationsError: 000004DC: LdapErr: DSID-0C090A22, comment: In order to perform this operation a successful bind must be completed on the connection., data 0, v3839

Or I would get Invalid username/password.

So, following some code examples I found elsewhere, I was able to change the code shown in the example to the following and it now works for me.

**default.json**
```
  "authentication": {
    "strategies": [
      "jwt",
      "local",
      "ldap"
    ],
    "local": {
      "entity": "user",
      "usernameField": "username",
      "passwordField": "password"
    },
    "ldap": {
      "name": "ldap",
      "server": {
        "url": "ldap://10.0.0.34:389/",
        "searchBase": "dc=mydomain,dc=com",
        "searchFilter": "(|(userPrincipalName={{username}})(sAMAccountName={{username}}))",
        "searchAttributes": null
      },
      "passRequestToCallback": true
    }
  },
```

**authentication.js**
```
const authentication = require('@feathersjs/authentication');
const jwt = require('@feathersjs/authentication-jwt');
const local = require('@feathersjs/authentication-local');
const ldap = require('feathers-authentication-ldap');

module.exports = function (app) {
  const config = app.get('authentication');

  // Set up authentication with the secret
  app.configure(authentication(config));
  app.configure(jwt());
  app.configure(local());
  app.configure(ldap({
    asyncOptions: req => {
      // If their username doesn't contain @domain.whatever, we need to add it
      // for Active Directory to be able to authenticate the user and look them up.
      // Grab it from the searchBase which looks like "searchBase": "cn=users,dc=mydomain,dc=com"
      let username = req.body.username || '';
      if (username.indexOf('@') === -1) {
        let domain = '';
        let searchBase = (config.ldap.server.searchBase || '').split(',');
        if (searchBase && searchBase.length) {
          searchBase.forEach(element => {
            if (element.includes('dc=')) {
              const indx = element.indexOf('=') + 1;
              if (domain) {
                domain += '.';
              }
              domain += element.substring(indx);
            }
          });
        }
        username += '@' + domain;
      }
      return({ server: { bindDn: username, bindCredentials: req.body.password } });
    }
  }));

  // The `authentication` service is used to create a JWT.
  // The before `create` hook registers strategies that can be used
  // to create a new valid JWT (e.g. local or oauth2)
  app.service('authentication').hooks({
    before: {
      create: [
        authentication.hooks.authenticate(config.strategies),
        function (hook) {
          // Restore the original authentication type from the jwt if they logged back in using only the jwt
          if (hook.data.strategy === 'jwt' && hook.data.accessToken) {
            const myjwt = decode(hook.data.accessToken);
            const strategy = myjwt.strategy;
            hook.params.payload = hook.params.payload || {};
            // hook.params.payload.userId should already be on the payload from the framework if using Local logins
            hook.params.payload.strategy = strategy;  // Store the auth strategy this JWT was issued against.
          } else {
            hook.params.payload.strategy = hook.data.strategy;
            // user.username when local auth, user.sAMAccountName when active directory
            hook.params.payload.username = hook.params.user.username || hook.params.user.sAMAccountName || '';
            hook.params.payload.userId = hook.params.payload.userId || null; // Or unique ID from whatever the user table is
          }
          return Promise.resolve(hook);
        }        
      ],
      remove: [
        authentication.hooks.authenticate('jwt')
      ]
    }
  });
};
```

**Postman API data**
```
// My Active Directory Usernames are firstname.lastname
// This works
{
	"strategy": "ldap",
	"username": "firstname.lastname",
	"password": "supersecret"
}

// And so does this
{
	"strategy": "ldap",
	"username": "firstname.lastname@mydomain.com",
	"password": "supersecret"
}
```

Hope this helps someone else struggling with Active Directory Login.

**Comment by aditya02maurya** [comment]
@nathanbrizzee did you use MAC or WIN for this?

**Comment by nathanbrizzee** [comment]
@aditya02maurya  I used Ubuntu Linux 18.04.  I updated all my packages to the latest versions today and now I get an error.  I tried it both on Windows 10 and Ubuntu Linux 18.04.  They both give the following error:

`
info: Feathers application started on http://localhost:3030
error: Unhandled Rejection at: Promise  {"_bitField":18087936,"_fulfillmentHandler0":{"name":"SequelizeAccessDeniedError","parent":{"message":"Login failed. The login is from an untrusted domain and cannot be used with Windows authentication.","code":"ELOGIN"},"original":{"message":"Login failed. The login is from an untrusted domain and cannot be used with Windows authentication.","code":"ELOGIN"}},"meta":[{"isFulfilled":false,"isRejected":true,"rejectionReason":{"name":"SequelizeAccessDeniedError","parent":{"message":"Login failed. The login is from an untrusted domain and cannot be used with Windows authentication.","code":"ELOGIN"},"original":{"message":"Login failed. The login is from an untrusted domain and cannot be used with Windows authentication.","code":"ELOGIN"}}},{"name":"SequelizeAccessDeniedError","parent":{"message":"Login failed. The login is from an untrusted domain and cannot be used with Windows authentication.","code":"ELOGIN"},"original":{"message":"Login failed. The login is from an untrusted domain and cannot be used with Windows authentication.","code":"ELOGIN"}}]}
`

I'm going to investigate it now since it broke.  I need it working for my application.
I'll post an update once I get it working again. :-)

**Comment by nathanbrizzee** [comment]
@aditya02maurya  I was using Ubuntu 18.04.  I just checked out my project and updated my package references to the latest and tried my code again both in Ubuntu 18.04 and Windows 7 and it still works.  If you are getting an error message, what does it say? I've found Active Directory can be pretty tricky and dependent upon how the System Administrator has configured it.

In authentication.js, where hook.params.payload.username is assigned in the else clause, try putting a console.log(hook.params.user); just before and see if anything shows up.
